src/functional.rs

```diff
@@ -84,24 +84,35 @@ impl<K, V> Tree<K, V> {
         match self {
             Tree::Leaf => Tree::Node(Node::new(key, value)),
             Tree::Node(n) => match key.cmp(&n.key) {
-                cmp::Ordering::Less => Tree::Node(Node {
-                    left: Rc::new(n.left.insert(key, value)),
-                    key: Rc::clone(&n.key),
-                    right: Rc::clone(&n.right),
-                    value: Rc::clone(&n.value),
-                }),
+                cmp::Ordering::Less => {
+                    let new_left = n.left.insert(key, value);
+                    let height = new_left.height().max(n.right.height()) + 1;
+                    Tree::Node(Node {
+                        height,
+                        key: Rc::clone(&n.key),
+                        left: Rc::new(new_left),
+                        right: Rc::clone(&n.right),
+                        value: Rc::clone(&n.value),
+                    })
```

Comment:
This pattern is very repetitive - I wonder if there's some cool thing we can do here. I'm not exactly sure what it'd be ... like `from_node` or something? But we have to be able to give it either a new left child or a new right child and I'm not exactly sure how that API won't be gross. We could probably double down on it taking a parameter of:
```rust
enum NewChild {
    Left(Node),
    Right(Node),
}
```
which seems overkill but since this is in here 4 times or something maybe it's worth it ⚖️

Reply:
Maybe the function should just take key, value, left child, right child and calculate the height.

Reply:
Or maybe `node.new_from_self(left_child, right_child)`